bsim: Bluetooth: mesh: fix model extension

HaavardRei opened the pull request:
In 'test_access.c', model 3 on element 1 extending model 3 on element 0 is now registered correctly. When calling 'bt_mesh_model_extend' on models on different elements, this needs to called from the model with the highest element index to be registered correctly.

HaavardRei committed:
bsim: Bluetooth: mesh: fix model extension

In 'test_access.c', model 3 on element 1 extending model 3 on element 0
is now registered correctly. When calling 'bt_mesh_model_extend' on
models on different elements, this needs to called from the
model with the highest element index to be registered correctly.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

alxelax commented:
Hm... I didn't find described in the comment limitation. Looking at the extension functionality implementation, this is just a linked list. It doesn't have dependency in the elements and be called for model from any element.

HaavardRei commented:
@alxelax I agree that the documentation does not specify that this is how it should be done. However, the current implementation results in this:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/91412738/6696276c-8f1f-4827-b30f-8c712f21ac75)
And the changes I proposed results in this:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/91412738/ba34ba33-8c1a-46e4-98fc-197732e2b1de)

The problem seems to be that the model is not yet initialized, resulting in (the case of the current implementation) the cfg client extending model 3.

alxelax reviewed (changes requested):
@HaavardRei, seems this dependency has been added with Composition Data page 1 implementation. 
This changed public API behavior without mentioning this dependency in API description.
For correct fixing of this, the fix only bsim test is not sufficient. It requires in fixing of the API description.

HaavardRei committed:
Bluetooth: mesh: update model extension

Specifies that if Composition Data Page 1 is enabled, the models passed
to 'bt_mesh_model_extend' needs to be initialized prior to the call.
This is to ensure that the composition data is registered correctly.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei commented:
@alxelax Added a new commit updating the API description of `bt_mesh_model_extend`

alxelax commented on include/zephyr/bluetooth/mesh/access.h:
```diff
@@ -1070,6 +1070,11 @@ void bt_mesh_model_data_store_schedule(struct bt_mesh_model *mod);
  *  extension list and element, giving the models extended subscription list
  *  capacity.
  *
+ * If @kconfig{CONFIG_BT_MESH_COMP_PAGE_1} is enabled, either base_mod or
```
with reference on CDP1 feature? Since without this, it can be called.

PavelVPV replied:
Ok, fine, leave CDP1 as a reference. This will become default later anyway.

HaavardRei replied:
Since the extension normally is a part of one of the init callbacks, perhaps:

>  It is not allowed to call this function before the :c:member:`bt_mesh_model_cb.init` callback is called for both models, or it can be called as part of the final callback.

HaavardRei replied:
Updated!
